lib/allowed_ops.ml

```diff
@@ -0,0 +1,383 @@
+open Xen_api_lwt_unix
+open Context
+
+type 'a api = rpc:rpc -> session_id:API.ref_session -> 'a
+
+module type S = sig
+  val name : string
+
+  val execute : Context.t -> unit Lwt.t
+end
+
+module type BEHAVIOUR = sig
+  type t
+
+  type operation
+
+  val rpc_of_operation : operation -> Rpc.t
+
+  val name : string
+
+  val get_uuid : (self:t -> string Lwt.t) api
+
+  val get_allowed_operations : (self:t -> operation list Lwt.t) api
+
+  val get_all : t list Lwt.t api
+
+  val perform : Context.session -> t -> operation -> unit Lwt.t
+end
+
+let on_self ctx self op = rpc ctx (fun ~rpc ~session_id -> op ~rpc ~session_id ~self)
+
+module Make (B : BEHAVIOUR) : S = struct
+  let name = B.name
+
+  let pp_operation = Fmt.using B.rpc_of_operation PP.rpc_t
+
+  let execute t =
+    step t B.name
+    @@ fun ctx ->
+    let seen = Hashtbl.create 17 in
+    let rec perform_allowed self =
+      Lwt.catch
+        (fun () ->
+          on_self ctx self B.get_uuid
+          >>= fun uuid ->
+          on_self ctx self B.get_allowed_operations
+          >>= fun ops ->
+          debug (fun m -> m "Available operations on %s: %a" uuid Fmt.(list pp_operation) ops) ;
+          match List.find_all (fun e -> not (Hashtbl.mem seen e)) ops with
+          | [] -> Lwt.return_unit
+          | op :: _ ->
+              Hashtbl.add seen op () ;
+              debug (fun m -> m "Performing %a on %s" pp_operation op uuid) ;
+              Lwt.catch
+                (fun () -> B.perform ctx self op)
+                (function
+                    | Api_errors.Server_error (code, (msg :: _ as lst))
+                      when code = Api_errors.sr_backend_failure && msg = "NotImplementedError" ->
+                        warn (fun m ->
+                            m "Operation %a is not implemented: %a!" pp_operation op
+                              Fmt.(list string)
+                              lst ) ;
+                        Lwt.return_unit
+                    | e ->
+                        err (fun m -> m "Operation %a failed: %a" pp_operation op Fmt.exn e) ;
+                        Lwt.return_unit
+                        (* and keep going *))
+              >>= fun () -> perform_allowed self )
+        (function
+            | Api_errors.Server_error (code, _) when code = Api_errors.handle_invalid ->
+                (* fine, we've run destroy/forget *)
+                Lwt.return_unit
+            | e -> Lwt.fail e)
+    in
+    rpc ctx B.get_all
+    >>= fun all ->
+    debug (fun m -> m "Performing operations serially") ;
+    Lwt_list.iter_s perform_allowed all
+    >>= fun () ->
+    rpc ctx B.get_all
+    >>= fun all ->
+    debug (fun m -> m "Performing operations in parallel") ;
```

Comment:
So this relies on the fact that we usually wait in message forwarding for like 20 mins if there is another operation in progress instead of failing?